rowspan="0" results in different table layout than Firefox/Chrome

karlcow opened the pull request:
#### 75a5507d4d8f004d43076ca036c32d4c798daccd
<pre>
rowspan=&quot;0&quot; results in different table layout than Firefox/Chrome
<a href="https://bugs.webkit.org/show_bug.cgi?id=185341">https://bugs.webkit.org/show_bug.cgi?id=185341</a>
<a href="https://rdar.apple.com/133910430">rdar://133910430</a>

Reviewed by Alan Baradlay.

<a href="https://html.spec.whatwg.org/multipage/tables.html#attr-tdth-rowspan">https://html.spec.whatwg.org/multipage/tables.html#attr-tdth-rowspan</a>
&gt; The td and th elements may also have a rowspan content attribute specified,
&gt; whose value must be
&gt; a valid non-negative integer less than or equal to 65534.
&gt; For this attribute, the value zero means that the cell is
&gt; to span all the remaining rows in the row group.

Basically is the equivalent of an imaginary all keyword.

&gt; The rowSpan IDL attribute must reflect the rowspan content attribute.
&gt; It is clamped to the range [0, 65534],
&gt; and its default value is 1.

We can put the value to the maximum number of rowspan, because
it will be clamped later on for the layout to the actual number of rows.

Setting rowspan=&quot;0&quot; is like setting a 3 or more on a table of 3 rows.

* LayoutTests/TestExpectations:
* LayoutTests/platform/glib/tables/mozilla_expected_failures/bugs/bug9879-1-expected.txt:
* LayoutTests/platform/glib/tables/mozilla/bugs/bug30332-1-expected.txt:
* LayoutTests/platform/glib/tables/mozilla/bugs/bug30332-2-expected.txt:
* LayoutTests/platform/glib/tables/mozilla/bugs/bug9879-1-expected.txt:
* LayoutTests/platform/ios/tables/mozilla_expected_failures/bugs/bug9879-1-expected.txt:
* LayoutTests/platform/ios/tables/mozilla/bugs/bug30332-1-expected.txt:
* LayoutTests/platform/ios/tables/mozilla/bugs/bug30332-2-expected.txt:
* LayoutTests/platform/ios/tables/mozilla/bugs/bug9879-1-expected.txt:
* LayoutTests/platform/mac/tables/mozilla_expected_failures/bugs/bug9879-1-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug30332-1-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug30332-2-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug9879-1-expected.txt:
* Source/WebCore/html/HTMLTableCellElement.cpp:
(WebCore::HTMLTableCellElement::rowSpan const):

Canonical link: <a href="https://commits.webkit.org/288746@main">https://commits.webkit.org/288746@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/44c06fce861cb45a54f086565bc8febfafe2b7da

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84305 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3928 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/38610 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/89383 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/35314 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/86390 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4014 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/11903 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/89383 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/35314 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87351 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/3011 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/76599 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/89383 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/2962 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/30829 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/34363 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/73859 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/31597 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/90764 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/11572 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/8404 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/90764 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11799 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/72424 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/90764 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/17542 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/15986 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/2937 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13045 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/11524 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/17000 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11373 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14849 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/13146 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

karlcow commented on Source/WebCore/html/HTMLTableCellElement.cpp:
```diff
@@ -59,7 +59,10 @@ unsigned HTMLTableCellElement::colSpan() const
-    // FIXME: a rowSpan equal to 0 should be allowed, and mean that the cell is to span all the remaining rows in the row group.
+    unsigned rowSpanValue = rowSpanForBindings();
+    // when rowspan=0, the specs says it should apply to the full rows.
+    if (rowSpanValue == 0)
```
How do I fix this?

> 'Tests for true/false, null/non-null, and zero/non-zero should all be done without equality comparisons.'

The stylechecker is not happy. But I found many places in the code with the same syntax.

Ahmad-S792 replied:
`if (!rowSpanValue)` works? (Actually never mind - we want to apply when it is '0').

twilco replied:
```cpp
if (!rowSpanValue)
    return maxRowSpan;
```

Should work. And maybe we should reuse `rowSpanValue` instead of calling `rowSpanForBindings()` again in `return std::max(1u, rowSpanForBindings());`?

karlcow replied:
Done! Thanks a lot @Ahmad-S792 and @twilco

nt1m commented on Source/WebCore/html/HTMLTableCellElement.cpp:
```diff
@@ -59,8 +59,11 @@ unsigned HTMLTableCellElement::colSpan() const
+    unsigned rowSpanValue = rowSpanForBindings();
+    // when rowspan=0, the specs says it should apply to the full rows.
+    if (!rowSpanValue)
+        return maxRowspan;
```
Chrome returns 1 in this case: https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/html/html_table_cell_element.cc;l=55?q=HTMLTableCellElement::rowSpan&ss=chromium

karlcow replied:
I'm not sure what Chrome is doing after with this one. because `rowspan="0"` is **not** the equivalent of `rowspan="1"`

![Screenshot 2025-01-07 at 13 31 52](https://github.com/user-attachments/assets/118d4d20-bee5-4a48-a642-f4ea8a2731d7)

karlcow replied:
@nt1m they do something similar in fact. 
Just found.
https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/layout/table/layout_table_cell.h;l=29-32;drc=f39c57f31413abcb41d3068cfb2c7a1718003cc5;bpv=1;bpt=1

alanbaradlay commented on Source/WebCore/html/HTMLTableCellElement.cpp:
```diff
@@ -59,8 +59,11 @@ unsigned HTMLTableCellElement::colSpan() const
-    // FIXME: a rowSpan equal to 0 should be allowed, and mean that the cell is to span all the remaining rows in the row group.
-    return std::max(1u, rowSpanForBindings());
+    unsigned rowSpanValue = rowSpanForBindings();
+    // when rowspan=0, the specs says it should apply to the full rows.
```
Could you please be a bit more specific about the 'specs' here? (I know you linked to it in the commit message, but I think this comment may read better with a tiny bit more context)

karlcow replied:
@alanbaradlay yes sure. Would that work?

```suggestion
    // when rowspan=0, the HTML spec says it should apply to the full remaining rows.
    // In https://html.spec.whatwg.org/multipage/tables.html#attr-tdth-rowspan
    // > For this attribute, the value zero means that the cell is
    // > to span all the remaining rows in the row group.
```

karlcow replied:
done. (sorry for the delay I was on sick leave)

karlcow committed:
rowspan="0" results in different table layout than Firefox/Chrome
https://bugs.webkit.org/show_bug.cgi?id=185341
rdar://133910430

Reviewed by Alan Baradlay.

https://html.spec.whatwg.org/multipage/tables.html#attr-tdth-rowspan
> The td and th elements may also have a rowspan content attribute specified,
> whose value must be
> a valid non-negative integer less than or equal to 65534.
> For this attribute, the value zero means that the cell is
> to span all the remaining rows in the row group.

Basically is the equivalent of an imaginary all keyword.

> The rowSpan IDL attribute must reflect the rowspan content attribute.
> It is clamped to the range [0, 65534],
> and its default value is 1.

We can put the value to the maximum number of rowspan, because
it will be clamped later on for the layout to the actual number of rows.

Setting rowspan="0" is like setting a 3 or more on a table of 3 rows.

* LayoutTests/TestExpectations:
* LayoutTests/platform/glib/tables/mozilla_expected_failures/bugs/bug9879-1-expected.txt:
* LayoutTests/platform/glib/tables/mozilla/bugs/bug30332-1-expected.txt:
* LayoutTests/platform/glib/tables/mozilla/bugs/bug30332-2-expected.txt:
* LayoutTests/platform/glib/tables/mozilla/bugs/bug9879-1-expected.txt:
* LayoutTests/platform/ios/tables/mozilla_expected_failures/bugs/bug9879-1-expected.txt:
* LayoutTests/platform/ios/tables/mozilla/bugs/bug30332-1-expected.txt:
* LayoutTests/platform/ios/tables/mozilla/bugs/bug30332-2-expected.txt:
* LayoutTests/platform/ios/tables/mozilla/bugs/bug9879-1-expected.txt:
* LayoutTests/platform/mac/tables/mozilla_expected_failures/bugs/bug9879-1-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug30332-1-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug30332-2-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug9879-1-expected.txt:
* Source/WebCore/html/HTMLTableCellElement.cpp:
(WebCore::HTMLTableCellElement::rowSpan const):

Canonical link: https://commits.webkit.org/288746@main
